frame/contracts/src/lib.rs

```diff
@@ -1008,19 +999,94 @@ impl Default for Schedule {
 	}
 }
 
+#[doc(hidden)]
+#[cfg_attr(test, derive(Debug))]
+pub struct DynamicWeightData<AccountId, NegativeImbalance> {
+	/// The account id who should receive the refund from the gas leftovers.
+	transactor: AccountId,
+	/// The negative imbalance obtained by withdrawing the value up to the requested gas limit.
+	imbalance: NegativeImbalance,
+}
+
 /// `SignedExtension` that checks if a transaction would exhausts the block gas limit.
 #[derive(Encode, Decode, Clone, Eq, PartialEq)]
 pub struct CheckBlockGasLimit<T: Trait + Send + Sync>(PhantomData<T>);
 
+impl<T: Trait + Send + Sync> CheckBlockGasLimit<T> {
+	/// Perform pre-dispatch checks for the given call and origin.
+	fn perform_pre_dispatch_checks(
+		who: &T::AccountId,
+		call: &<T as Trait>::Call,
+	) -> Result<Option<DynamicWeightData<T::AccountId, NegativeImbalanceOf<T>>>, TransactionValidityError> {
+		let call = match call.is_sub_type() {
+			Some(call) => call,
+			None => return Ok(None),
+		};
+
+		match *call {
+			Call::claim_surcharge(_, _) | Call::update_schedule(_) | Call::put_code(_) => Ok(None),
+			Call::call(_, _, gas_limit, _) | Call::instantiate(_, gas_limit, _, _) => {
+				// Compute how much block weight this transaction can take at its limit, i.e.
+				// if this transaction depleted all provided gas to zero.
+				let gas_weight_limit = Weight::try_from(gas_limit)
+					.map_err(|_| InvalidTransaction::ExhaustsResources)?;
+				let weight_available = <T as system::Trait>::MaximumBlockWeight::get()
+					.saturating_sub(<system::Module<T>>::all_extrinsics_weight()).into();
+				if gas_weight_limit > weight_available {
+					// We discard the transaction if the requested limit exceeds the available
+					// amount of weight in the current block.
+					//
+					// Note that this transaction is left out only from the current block and txq
+					// might attempt to include this transaction again.
+					return Err(InvalidTransaction::ExhaustsResources.into());
+				}
+
+				// Compute the fee corresponding for the given gas_weight_limit and attempt
+				// withdrawing from the origin of this transaction.
+				let fee = T::WeightToFee::convert(gas_weight_limit);
+
+				// Compute and store the effective price per unit of gas.
+				let gas_price = {
```

Comment:
This seems problematic, as it is assuming a linear weight-to-fee conversion, which may be not be the case.